### PR TITLE
Live Queries Support.

### PR DESCRIPTION
**Pull request by peterdotjs**
cc: @wangmengyan95

### CODE REVIEW

File: src/CoreManager.js [hunk 1]
```diff
@@ -455,5 +456,25 @@ module.exports = {
...
+      throw new Error('LiveQueryController must implement subscribe()');
+    }
+    if (typeof controller.unsubscribe !== 'function') {
+      throw new Error('LiveQueryController must implement unsubscribe()');
+    }
+    if (typeof controller.open !== 'function') {
+      throw new Error('LiveQueryController must implement open()');
+    }
+    if (typeof controller.close !== 'function') {
+      throw new Error('LiveQueryController must implement disconnect()');
```
Comment (wangmengyan95) [inline src/CoreManager.js]: Nit: change to `LiveQueryController must implement close()`

File: src/LiveQueryClient.js [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+  }
+
+  /**
+   * Subscribes to a ParseQuery
+   * 
+   * If you provide the sessionToken, when the LiveQuery server gets ParseObject's 
+   * updates from parse server, it'll try to check whether the sessionToken fulfills 
+   * the ParseObject's ACL. The LiveQuery server will only send updates to clients whose 
+   * sessionToken is fit for the ParseObject's ACL. You can check the LiveQuery protocol
+   * here for more details. The subscription you get is the same subscription you get
```
Comment (wangmengyan95) [inline src/LiveQueryClient.js]: Provide link for here

File: src/LiveQueryClient.js [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+  /**
+   * Subscribes to a ParseQuery
+   * 
+   * If you provide the sessionToken, when the LiveQuery server gets ParseObject's 
+   * updates from parse server, it'll try to check whether the sessionToken fulfills 
+   * the ParseObject's ACL. The LiveQuery server will only send updates to clients whose 
+   * sessionToken is fit for the ParseObject's ACL. You can check the LiveQuery protocol
+   * here for more details. The subscription you get is the same subscription you get 
+   * from our Standard API. You can check our Standard API about how to use the 
+   * subscription to get events.
```
Comment (wangmengyan95) [inline src/LiveQueryClient.js]: Not sure whether the last setence is useful.

File: src/LiveQuerySubscription.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+import events from 'events';
+import CoreManager from './CoreManager';
+
+/**
+ * Creates a new Live Query Subscription.
```
Comment (wangmengyan95) [inline src/LiveQuerySubscription.js]: `Live Query` to `LiveQuery`

File: src/LiveQuerySubscription.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+ * get this event. The object is the ParseObject which is deleted.
+ * 
+ * <pre>
+ * subscription.on('delete', (object) => {
+ * 
+ * });</pre></p>
+ *
+ *
+ * <p>Close Event - When the client loses the WebSocket connection to the LiveQuery
+ * server and we can't get anymore events, you will get this event.
```
Comment (wangmengyan95) [inline src/LiveQuerySubscription.js]: `get events anymore`

File: src/ParseLiveQuery.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    // If we can not find Parse.liveQueryServerURL, we try to extract it from Parse.serverURL
+    if (!liveQueryServerURL) {
+      let host = url.parse(CoreManager.get('SERVER_URL')).host;
+      liveQueryServerURL = 'ws://' + host;
+      CoreManager.set('LIVEQUERY_SERVER_URL', liveQueryServerURL);
+    }
+    // If liveQueryServerURL is not set, throw exception
+    if (!liveQueryServerURL) {
+      throw new Error('You need to set Parse.liveQueryServerURL before using LiveQuery');
+    }
```
Comment (wangmengyan95) [inline src/ParseLiveQuery.js]: Where is the `ws` protocol check, I think it should be here right?

Comment (peterdotjs) [inline src/ParseLiveQuery.js]: the check is here: https://github.com/ParsePlatform/Parse-SDK-JS/blob/peterjs.liveQuery/src/LiveQueryClient.js#L246

**Comment by peterdotjs** [comment]
Updated per comments.

**Comment by andrewimm** [comment]
This all looks good to me. I'll do one last pass, then merge into the 1.8 release branch, followed by merging that into master. If the rye tests I just landed all pass, I'll cut a second RC on `npm`, which we can test as a standalone package. If that all looks good, we can bump the version number to 1.8.0 and push that out to everyone on npm.

File: src/ParseLiveQuery.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+    let liveQueryServerURL = CoreManager.get('LIVEQUERY_SERVER_URL');
+    
+    if (liveQueryServerURL && liveQueryServerURL.indexOf('ws') !== 0) {
+      throw new Error('You need to set a proper Parse LiveQuery server url before using LiveQueryClient');
+    }
+
+    // If we can not find Parse.liveQueryServerURL, we try to extract it from Parse.serverURL
+    if (!liveQueryServerURL) {
+      let host = url.parse(CoreManager.get('SERVER_URL')).host;
```
Comment (andrewimm) [inline src/ParseLiveQuery.js]: This should really be a regex. There's no need to import the entire `url` browserify module, it's too much overhead (I know we've discussed this at least once before :).
It seems like what we really want is something like:
```js
let host = CoreManager.get('SERVER_URL').replace(/^https?:\/\//, '');
```

**Comment by andrewimm** [comment]
Made the changes, everything is merged. You can fetch the latest version of 1.8.0 with `npm install parse@1.8.0-rc2`

File: src/ParseLiveQuery.js (hunk 7 above)
Comment (peterdotjs) [inline src/ParseLiveQuery.js]: Yeah I remember. I didn't write this part but I did miss it in the review so good catch. The url package wasn't even in package.json so this part was a miss overall.

**Pull request merged**